### ISSUE TITLE
Integrate with OCaml language server

### ISSUE THREAD

**Issue by ghost** [reasonml-editor/reason-mode#5]
I recently tried the OCaml language server with Emacs and added some brief instructions on how to configure it [here](https://github.com/freebroccolo/ocaml-language-server#emacs). This is using [lsp-mode](https://github.com/emacs-lsp/lsp-mode).

It might be worth including the `lsp-mode` integration in this plugin directly and shipping it already set up to communicate with the language server.

I have not tested the server extensively with `lsp-mode` but it does at least appear to work to a minimal degree for things like flycheck, etc. In principle, most of the other features should work too.

If you want to give it a try, make sure to build the server off the master branch since it contains some fixes for editors like emacs that I have not published in a release version just yet. Planning to do that very soon.

**Comment by arichiardi** [comment]
I am new to this and I will check the links, but what are the main advantages over `merlin`?
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by ghost** [comment]
The OCaml Language Server is basically a wrapper around `merlin`. The main advantage is that is speaks the [Language Server Protocol](http://langserver.org/) which originated with VS Code. The OCaml Language Server is what I use as the backend for [vscode-reasonml](https://github.com/freebroccolo/vscode-reasonml).

The main idea is that if you have a language server that speaks the common LSP protocol, you should be able to reuse a common framework for handling a lot of the rich editing features like error reporting, hover details, etc., without having to do the whole thing from scratch. That's what [lsp-mode](https://github.com/emacs-lsp/lsp-mode) provides.

So basically the OCaml Language Server gives you the same features (in theory) as the VS Code plugin. You just need a bit of glue from the `lsp-mode` package to expose those features to Emacs.

**Comment by arichiardi** [comment]
Got it, I am not against it but I will prioritize the outstanding PR and Travis+Elpa setup for now. Flycheck integration would be awesome and on the roadmap, especially if coupled with BetterErrors (does `lsp-mode` provide that?).

**Comment by ghost** [comment]
Regarding error reporting, the language server protocol does provide functionality for this, along with something called "code actions" which are like error or warning messages along with options to perform a function or series of functions to fix an error or otherwise transform the code automatically (e.g., adding a missing semicolon or fixing a misspelled record field).

Currently the OCaml language server just uses the errors reported by Merlin in addition to some very basic code actions I have implemented. I have considered trying to integrate BetterErrors more directly though and may take a closer look at that soon. In theory, there is no reason why it couldn't be done because there is a lot of flexibility in how we process error reporting in the server before presenting it to the editors.
